File: packages/core/src/tool-result-codes.ts
/**
 * Tool-result code helpers — type-narrowed guards for the discriminated
 * `tool_result.output` union surfaced on streamed `ToolCallRecord`
 * values.
 *
 * The five variants are inert wire shapes — they only describe what the
 * backend may stamp on `tool_result.output`. The guards turn `unknown`
 * into one of the five `*Output` interfaces so callers can branch
 * without casting.
 *
 * Each guard checks both the `code` discriminant against
 * TOOL_RESULT_CODES AND its own required sibling fields — so a
 * well-formed `code` with a missing sibling returns false rather
 * than narrowing positive on the discriminant alone. The exhaustive-
 * match utility (`assertExhaustiveToolResult`) makes a switch over
 * ToolResultCode fail to compile if a sixth variant lands without
 * the consumer updating their handler.
 */

import type { ToolCallRecord } from "@codespar/types";

export const ToolResultCode = {
  PolicyDenied: "policy_denied",
  ApprovalRequired: "approval_required",
  MocksExhausted: "mocks_exhausted",
  MocksEngineError: "mocks_engine_error",
  ToolNotMocked: "tool_not_mocked",
} as const;

export type ToolResultCode = (typeof ToolResultCode)[keyof typeof ToolResultCode];

export const TOOL_RESULT_CODES: ReadonlySet<ToolResultCode> = new Set([
  ToolResultCode.PolicyDenied,
  ToolResultCode.ApprovalRequired,
  ToolResultCode.MocksExhausted,
  ToolResultCode.MocksEngineError,
  ToolResultCode.ToolNotMocked,
]);

export interface PolicyDeniedOutput {
  code: typeof ToolResultCode.PolicyDenied;
  rule_id: string;
  message: string;
}

export interface ApprovalRequiredOutput {
  code: typeof ToolResultCode.ApprovalRequired;
  approval_id: string;
  expires_at: string;
  message: string;
}

export interface MocksExhaustedOutput {
  code: typeof ToolResultCode.MocksExhausted;
  message: string;
}

export interface MocksEngineErrorOutput {
  code: typeof ToolResultCode.MocksEngineError;
  message: string;
}

export interface ToolNotMockedOutput {
  code: typeof ToolResultCode.ToolNotMocked;
  tool_name: string;
  message: string;
}

export type ToolResultOutcome =
  | PolicyDeniedOutput
  | ApprovalRequiredOutput
  | MocksExhaustedOutput
  | MocksEngineErrorOutput
  | ToolNotMockedOutput;

// Narrowed ToolCallRecord aliases — when a guard succeeds the
// `output` field is known to be the corresponding *Output variant.
export type PolicyDeniedToolCall = ToolCallRecord & {
  output: PolicyDeniedOutput;
};
export type ApprovalRequiredToolCall = ToolCallRecord & {
  output: ApprovalRequiredOutput;
};
export type MocksExhaustedToolCall = ToolCallRecord & {
  output: MocksExhaustedOutput;
};
export type MocksEngineErrorToolCall = ToolCallRecord & {
  output: MocksEngineErrorOutput;
};
export type ToolNotMockedToolCall = ToolCallRecord & {
  output: ToolNotMockedOutput;
};

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function readStringField(obj: Record<string, unknown>, key: string): string | null {
  const v = obj[key];
  return typeof v === "string" ? v : null;
}

export function isPolicyDenied(value: unknown): value is PolicyDeniedOutput {
  if (!isObject(value)) return false;
  if (value.code !== ToolResultCode.PolicyDenied) return false;
  if (!TOOL_RESULT_CODES.has(value.code as ToolResultCode)) return false;
  return readStringField(value, "rule_id") !== null
    && readStringField(value, "message") !== null;
}

export function isApprovalRequired(value: unknown): value is ApprovalRequiredOutput {
  if (!isObject(value)) return false;
  if (value.code !== ToolResultCode.ApprovalRequired) return false;
  if (!TOOL_RESULT_CODES.has(value.code as ToolResultCode)) return false;
  return readStringField(value, "approval_id") !== null
    && readStringField(value, "expires_at") !== null
    && readStringField(value, "message") !== null;
}

export function isMocksExhausted(value: unknown): value is MocksExhaustedOutput {
  if (!isObject(value)) return false;
  if (value.code !== ToolResultCode.MocksExhausted) return false;
  if (!TOOL_RESULT_CODES.has(value.code as ToolResultCode)) return false;
  return readStringField(value, "message") !== null;
}

export function isMocksEngineError(value: unknown): value is MocksEngineErrorOutput {
  if (!isObject(value)) return false;
  if (value.code !== ToolResultCode.MocksEngineError) return false;
  if (!TOOL_RESULT_CODES.has(value.code as ToolResultCode)) return false;
  return readStringField(value, "message") !== null;
}

export function isToolNotMocked(value: unknown): value is ToolNotMockedOutput {
  if (!isObject(value)) return false;
  if (value.code !== ToolResultCode.ToolNotMocked) return false;
  if (!TOOL_RESULT_CODES.has(value.code as ToolResultCode)) return false;
  return readStringField(value, "tool_name") !== null
    && readStringField(value, "message") !== null;
}

/**
 * Exhaustive-match witness. A `switch` over `ToolResultCode` should
 * pass `value` here in the default branch — TS fails to compile if
 * `value` isn't narrowed to `never`, i.e. if a code variant escaped
 * the switch. The runtime body throws so a hostile cast at runtime
 * doesn't silently swallow an unknown code.
 */
export function assertExhaustiveToolResult(value: never): never {
  throw new Error(
    `tool-result-codes: unexpected output variant ${JSON.stringify(value)}`,
  );
}

File: packages/python/src/codespar/tool_result_codes.py
"""
Tool-result code helpers — Python parallel of tool-result-codes.ts.

Mirrors the TypeScript surface 1:1: five frozen dataclasses, five
discriminant constants, the ``ToolResultCode`` Literal union, the
``TOOL_RESULT_CODES`` frozenset, five PEP 647 ``TypeGuard``
predicates, and ``assert_exhaustive_tool_result``.

Each guard checks both the ``code`` discriminant against
``TOOL_RESULT_CODES`` AND its own required sibling fields. A
payload with a well-formed ``code`` but a missing sibling returns
False rather than narrowing positive on the discriminant alone —
the test suite enforces this invariant per guard.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Final, Literal, TypeGuard

# ── discriminant constants ────────────────────────────────────────

POLICY_DENIED: Final[Literal["policy_denied"]] = "policy_denied"
APPROVAL_REQUIRED: Final[Literal["approval_required"]] = "approval_required"
MOCKS_EXHAUSTED: Final[Literal["mocks_exhausted"]] = "mocks_exhausted"
MOCKS_ENGINE_ERROR: Final[Literal["mocks_engine_error"]] = "mocks_engine_error"
TOOL_NOT_MOCKED: Final[Literal["tool_not_mocked"]] = "tool_not_mocked"

ToolResultCode = Literal[
    "policy_denied",
    "approval_required",
    "mocks_exhausted",
    "mocks_engine_error",
    "tool_not_mocked",
]

TOOL_RESULT_CODES: Final[frozenset[str]] = frozenset(
    [
        POLICY_DENIED,
        APPROVAL_REQUIRED,
        MOCKS_EXHAUSTED,
        MOCKS_ENGINE_ERROR,
        TOOL_NOT_MOCKED,
    ]
)


# ── output dataclasses (frozen, slotted) ──────────────────────────


@dataclass(slots=True, frozen=True)
class PolicyDeniedOutput:
    rule_id: str
    message: str
    code: Literal["policy_denied"] = POLICY_DENIED


@dataclass(slots=True, frozen=True)
class ApprovalRequiredOutput:
    approval_id: str
    expires_at: str
    message: str
    code: Literal["approval_required"] = APPROVAL_REQUIRED


@dataclass(slots=True, frozen=True)
class MocksExhaustedOutput:
    message: str
    code: Literal["mocks_exhausted"] = MOCKS_EXHAUSTED


@dataclass(slots=True, frozen=True)
class MocksEngineErrorOutput:
    message: str
    code: Literal["mocks_engine_error"] = MOCKS_ENGINE_ERROR


@dataclass(slots=True, frozen=True)
class ToolNotMockedOutput:
    tool_name: str
    message: str
    code: Literal["tool_not_mocked"] = TOOL_NOT_MOCKED


ToolResultOutcome = (
    PolicyDeniedOutput
    | ApprovalRequiredOutput
    | MocksExhaustedOutput
    | MocksEngineErrorOutput
    | ToolNotMockedOutput
)


# ── guards ────────────────────────────────────────────────────────


def _is_object(value: Any) -> TypeGuard[dict[str, Any]]:
    return isinstance(value, dict)


def _has_str(obj: dict[str, Any], key: str) -> bool:
    return isinstance(obj.get(key), str)


def is_policy_denied(value: Any) -> TypeGuard[dict[str, Any]]:
    if not _is_object(value):
        return False
    code = value.get("code")
    if code != POLICY_DENIED or code not in TOOL_RESULT_CODES:
        return False
    return _has_str(value, "rule_id") and _has_str(value, "message")


def is_approval_required(value: Any) -> TypeGuard[dict[str, Any]]:
    if not _is_object(value):
        return False
    code = value.get("code")
    if code != APPROVAL_REQUIRED or code not in TOOL_RESULT_CODES:
        return False
    return (
        _has_str(value, "approval_id")
        and _has_str(value, "expires_at")
        and _has_str(value, "message")
    )


def is_mocks_exhausted(value: Any) -> TypeGuard[dict[str, Any]]:
    if not _is_object(value):
        return False
    code = value.get("code")
    if code != MOCKS_EXHAUSTED or code not in TOOL_RESULT_CODES:
        return False
    return _has_str(value, "message")


def is_mocks_engine_error(value: Any) -> TypeGuard[dict[str, Any]]:
    if not _is_object(value):
        return False
    code = value.get("code")
    if code != MOCKS_ENGINE_ERROR or code not in TOOL_RESULT_CODES:
        return False
    return _has_str(value, "message")


def is_tool_not_mocked(value: Any) -> TypeGuard[dict[str, Any]]:
    if not _is_object(value):
        return False
    code = value.get("code")
    if code != TOOL_NOT_MOCKED or code not in TOOL_RESULT_CODES:
        return False
    return _has_str(value, "tool_name") and _has_str(value, "message")


def assert_exhaustive_tool_result(value: Any) -> None:
    """Raise on any tool-result payload not matched by the five guards.

    Equivalent to TypeScript's ``assertExhaustiveToolResult(value: never)``
    — call from a default branch after handling each variant so a sixth
    code landing without a handler trips at the boundary instead of
    being silently swallowed.
    """
    raise AssertionError(
        f"tool-result-codes: unexpected output variant {value!r}",
    )
